src/main/scala/org/constellation/domain/transaction/PendingTransactionsMemPool.scala

```diff
@@ -54,25 +53,31 @@ class PendingTransactionsMemPool[F[_]: Concurrent](
           case (addressHash, txs) =>
             transactionChainService
               .getLastAcceptedTransactionRef(addressHash)
-              .flatTap { last =>
-                logger.info(s"${Console.YELLOW}Last accepted: ${last} | Txs head: ${txs.headOption
-                  .map(_.transaction.lastTxRef)} hash=${txs.headOption.map(_.transaction.hash)}${Console.RESET}")
-              }
               .map { last =>
-                val ordinal = last.ordinal
-                val left = txs.filter(t => t.transaction.lastTxRef.ordinal >= ordinal || t.transaction.isTest) // TODO: get rid of >=
-                (last, left)
+                val testTxs = txs.filter(_.transaction.isTest)
+                val consecutiveTransactions = takeConsecutiveTransactions(last, txs)
+                val testTxsToInclude = testTxs diff consecutiveTransactions
+
+                (last, consecutiveTransactions ++ testTxsToInclude)
               }
-              .map {
-                case (last, t) =>
-                  if (t.headOption.map(_.transaction.lastTxRef).contains(last)) t
-                  else List.empty[TransactionCacheData]
+              .flatTap { case (last, consecTxs) =>
+                logger.info(s"${Console.YELLOW}Last accepted: ${last} | Txs head: ${consecTxs.headOption
+                  .map(_.transaction.lastTxRef)} hash=${consecTxs.headOption.map(_.transaction.hash)}${Console.RESET}")
               }
+              .map { case (_, consecTxs) => consecTxs }
         }
       }
       .map(
         _.sortBy(_.map(-_.transaction.fee.getOrElse(0L)).sum).flatten
       )
+
+  private def takeConsecutiveTransactions(prevLastTxRef: LastTransactionRef, txs: List[TransactionCacheData]): List[TransactionCacheData] = {
+    txs.find(_.transaction.lastTxRef == prevLastTxRef) match {
+      case Some(tx) =>
+        tx +: takeConsecutiveTransactions(LastTransactionRef(tx.transaction.hash, tx.transaction.ordinal), txs diff List(tx))
+      case None => Nil
+    }
+  }
```

Comment:
we need to be sure this recursion will not blow up. either by tailrec or making it iterative.

Reply:
Yeah, add `@tailrec` to function

Reply:
Sure will do that.